Feature #27 상품 상세 정보 읽기

sunggwon opened the pull request:
상품 상제 정보 읽기 기능 및 테스트 구현

sunggwon committed:
상품 상세 정보 읽기 기능 구현

sunggwon committed:
상품 상세 정보 읽기 기능 테스트 작성

sunggwon committed:
컨트롤러 쿼리 스트링 입력 실패 테스트 작성

f-lab-robert commented on src/main/java/com/flabedu/small/small/exception/CustomExceptionHandler.java:
```diff
@@ -21,6 +22,18 @@ public ResponseEntity<Object> handleInvalidValueException(MethodArgumentNotValid
     }
 
+    @ExceptionHandler(BindException.class)
+    public ResponseEntity<Object> handleGetException(BindException e){
```
- requestDTO 에 원하지 않은 값이 들어왔을 때를 대비하신걸까요 ? 그럴 경우에는 기존에 정의되어 있는 BAD REQUEST 를 활용해볼 수는 없을까요 ? CustomException 은 도메인 상에서의 에러를 api 사용자에게 제공하기 위함입니다.

sunggwon replied:
수정하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/mapper/ItemMapper.java:
```diff
@@ -12,14 +14,23 @@
     void addItemDetail(@Param("itemId") Long itemId, @Param("itemDetails") List<ItemDetail> itemDetails);
+
+    //Update
+    void setStock(@Param("detailId") long detailId, @Param("stock")long stock);
```
본 PR 에는 GET 에 관한 로직만 있어야 하지 않나요 ?

sunggwon replied:
가독성을 위해 코드 순서를 변경한 부분이 잡힌것 같습니다. 멘토링 받은데로 이와 같은 부분도 주의 하도록 하겠습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/model/CategoryInfo.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.*;
+
+@Data
+public class CategoryInfo {
```
model 이란 폴더는 이미 이전에 사용하지 않기로 했던거 같은데요

sunggwon replied:
폴더명을 dao로 수정하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/service/ItemService.java:
```diff
@@ -39,6 +41,53 @@ public void purchaseItem(OrderRequestDTO itemInfo, String userID){
     }
 
+    @Transactional
+    public void addItem(ItemRequestDTO newItem) {
```
본 PR title 에 해당하지 않는 로직인 것 같습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/controller/ItemController.java:
```diff
@@ -21,4 +22,9 @@ public ResponseEntity<Object> addItem(@RequestBody @Valid ItemRequestDTO newItem
         return ResponseEntity.status(HttpStatus.CREATED).build();
     }
+
+    @GetMapping("/product")
```
@GetMapping("/product/${id}")

@GetMapping("/product?id=1")

sunggwon replied:
Path 파라미터를 받는 형식으로 수정하였습니다.

sunggwon committed:
패키지 이름 model -> dao로 변경

ItemController에 getItems 메서드 시그니처 및 예외 처리방법 변경
  - ItemController.java
  - ItemControllerTest.java
  - CustomExceptionHandler.java

sunggwon committed:
SelectedItemRequestDTO 삭제

sunggwon committed:
불필요한 어노테이션 삭제

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/dto/response/SelectedItemResponseDTO.java:
```diff
@@ -0,0 +1,46 @@
+    private String itemNameEn;
+
+    @NotNull(message = "성별을 선택해주세요.")
+    private String gender;
```
성별은 Enum 으로 관리할 수 있지 않을까요

sunggwon replied:
성별 enum으로 처리하였습니다.
기존 DAO도 String으로 처리 되어 있는데 DAO를 변경하면 해당 이슈와 상관없는 코드들이 변경될거 같아서 DAO는 수정하지 않고 현재 PR에서는 DTO에서 String으로 변경되도록 처리하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/dto/response/SelectedItemResponseDTO.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Valid
+    @NotNull(message = "상품 사이즈 및 재고를 최소 1개 이상 입력해주세요.")
+    @Size(min = 1, message = "상품 사이즈 및 재고를 최소 1개 이상 입력해주세요.")
```
- 여기서 Valid 어노테이션이 필요한가요 ?
- Size 만으로 NotNull 의 효과를 볼 수는 없나요 ?

sunggwon replied:
불필요한 어노테이션 삭제하였습니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/controller/ItemControllerTest.java:
```diff
@@ -381,4 +384,41 @@ public void addItemItemDetailIsNegative() throws Exception {
+        ).andExpectAll(
+                status().isOk()
+        ).andDo(print());
+        Mockito.verify(itemService).getItem(id);
```
이 부분의 코드의 맥락이 어떻게 되는걸까요 ?

"ItemService의 getItem을 호출한다" 의 부분은 E2E 관점에서 왜 테스트가 필요한걸까요 ?
Input 이 있을 때 output 이 제대로 기대한 값이 나오는지가 중요한게 아닐까요 ?

sunggwon replied:
E2E 관점으로 테스트 코드를 작성하였고 입력과 출력이 명확하지 않는 코드들은 제거 하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/controller/ItemController.java:
```diff
@@ -21,4 +21,12 @@ public ResponseEntity<Object> addItem(@RequestBody @Valid ItemRequestDTO newItem
     }
+
+    @GetMapping("/product")
+    public ResponseEntity<Object> getItems(@RequestParam("id") long itemId){
```
path 형태가 아닌 RequestParam 형태를 사용하신 이유가 있을까요 ?

sunggwon replied:
Path 형태가 REST API 방식에 더 적합할거 같아 Path 형태로 변경하였습니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/controller/ItemControllerTest.java:
```diff
@@ -381,4 +384,41 @@ public void addItemItemDetailIsNegative() throws Exception {
+    }
+
+    @Test
+    @DisplayName("음수 ID를 0으로 입력하면 400 BAD_REQUEST 응답한다.")
```
음수 ID를 0으로 입력하면 

글의 수정이 필요해보입니다.

sunggwon replied:
수정하였습니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/mapper/ItemMapperTest.java:
```diff
@@ -1,24 +1,57 @@
-        itemMapper.findItemAll();
+        var list = itemMapper.findItemAll();
+
+        Assertions.assertNotNull(list);
```
BeforeEach 에 있는 내용들을 우선 본 데이터를 필요로 하는 함수로 이동시켜보시겠어요 ?
그리고

1. given

2. when

3. then 

주석과 함께 위의 형태로 작성해보시겠어요 ? 
나아가, input 값이 명확한 상황은 output (Then) 을 검증할 때, input 의 값이 맞는지를 확인하는게 더 정확할 것 같습니다. null 의 여부를 확인하는건 정확하지 않은 테스트 인것 같습니다.

sunggwon replied:
이슈와 관련있는 메서드만 처리하려다가 해당 테스트 파일의 모든 테스트 코드를 코멘트 해주신데로 변경하였습니다.
따라서 이슈와 관련없는 테스트 코드도 변경으로 잡힌거 같습니다.
리뷰 해주실 때 참고 부탁드립니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/mapper/ItemMapperTest.java:
```diff
@@ -68,4 +103,42 @@ public void setStockTest(){
         var item = itemMapper.findItemDetailById(detailId);
         Assertions.assertEquals(stock, item.getStock());
     }
+    //////////////////////////////////////////
```
함께 작성하는 코드인 만큼 이런 부분들은 지양해달라고 말씀드렸던 것 같습니다.

sunggwon replied:
삭제하였습니다.

sunggwon committed:
- 호출 방식 PathParam 형식으로 변경
- 불필요한 어노테이션 제거
- SelectedItemResponseDTO 성별 Enum 형식으로 변경
- 테스트 코드 given/when/then 형식으로 변경

sunggwon committed:
- ItemService ResponseDTO Enum 변경에 의한 처리

sunggwon committed:
- 충돌 처리

sunggwon committed:
- 충돌 처리

sunggwon committed:
- 충돌 처리

sunggwon committed:
- 충돌 처리

sunggwon committed:
Merge branch 'develop' into feature-#27

# Conflicts:
#	src/main/java/com/flabedu/small/small/mapper/ItemMapper.java
#	src/main/java/com/flabedu/small/small/service/ItemService.java
#	src/main/java/com/flabedu/small/small/web/dto/request/ItemDetailRequestDTO.java
#	src/test/java/com/flabedu/small/small/mapper/ItemMapperTest.java
#	src/test/java/com/flabedu/small/small/service/ItemServiceTest.java

sunggwon committed:
- 충돌 처리
